pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -872,8 +872,14 @@ public void ensureAllPartitionsConsuming(TableConfig tableConfig, PartitionLevel
       if (idealState.isEnabled()) {
         List<PartitionGroupConsumptionStatus> currentPartitionGroupConsumptionStatusList =
             getPartitionGroupConsumptionStatusList(idealState, streamConfig);
+
+        // Read the smallest offset when a new partition is detected
+        OffsetCriteria originalOffsetCriteria = streamConfig.getOffsetCriteria();
+        streamConfig.setOffsetCriteria(OffsetCriteria.SMALLEST_OFFSET_CRITERIA);
```

Comment:
Is the table config being pulled from cache? If so, will we be modifying the streamconfig of that tableconfig? Or, if someone re-uses it down the road and gets the wrong value in different part of the code.

I think a better way may be to update the setupNewPartitionGroup() method to take an additional boolean arg isLiveTable. If it is a live table, then pick smallest offset criteria

Reply:
Good question. When creating the `StreamConfig`, the map from the table config is copied, and maintained separately, so there is no race condition on that.

I actually started with the suggested approach by adding OffsetCriteria override to the method, then found that I need to carry the extra parameter all the way to the `StreamMetadataProvider` interface, and could pollute the interface and potentially cause unexpected behavior for the implementations of the interface. Then I decided to go with the changing stream config approach.

Reply:
I think we can use the local private method to get the earliest offset. These should be very rare, so an extra call to the stream should not matter.

I would prefer that we not change the table config in any way in case someone adds some code in this flow to pick up the table config and make some decision based on it.